File: Samples/MapScenariosObjC/MapScenariosObjC/Scenarios/RoutingViewController.m
//
//  RoutingViewController.m
//  MapScenariosObjC
//
//  Created on 3/7/18.
//  Copyright © 2018 Phunware. All rights reserved.
//

#import <PWMapKit/PWMapKit.h>
#import <PWCore/PWCore.h>

#import "RoutingViewController.h"

@interface RoutingViewController () <PWMapViewDelegate, CLLocationManagerDelegate>

@property (nonatomic, strong) PWMapView *mapView;
@property (nonatomic, strong) CLLocationManager *locationManager;
@property (nonatomic, assign) BOOL firstLocationAcquired;

@end

@implementation RoutingViewController

- (id)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        _buildingIdentifier = 0; // Enter your building identifier here, found on the building's Edit page on Maas portal
        // Enter your application identifier, access key, and signature key, found on Maas portal under Account > Apps
        _applicationId = @"";
        _accessKey = @"";
        _signatureKey = @"";
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"Route to Point of Interest";
    
    if (self.applicationId.length > 0 && self.accessKey.length > 0 && self.signatureKey.length > 0 && self.buildingIdentifier != 0) {
        [PWCore setApplicationID:self.applicationId accessKey:self.accessKey signatureKey:self.signatureKey];
    } else {
        [NSException raise:@"MissingConfiguration" format:@"applicationId, accessKey, signatureKey, and buildingIdentifier must be set"];
    }
    
    self.locationManager = [CLLocationManager new];
    self.locationManager.delegate = self;
    [self.locationManager requestWhenInUseAuthorization];
    
    self.mapView = [PWMapView new];
    self.mapView.delegate = self;
    [self.view addSubview:self.mapView];
    [self configureMapViewConstraints];
    
    [PWBuilding buildingWithIdentifier:self.buildingIdentifier completion:^(PWBuilding *building, NSError *error) {
        __weak typeof(self) weakSelf = self;
        [weakSelf.mapView setBuilding:building animated:YES onCompletion:^(NSError *error) {
            [weakSelf startManagedLocationManager];
        }];
    }];
}

- (void)startManagedLocationManager {
    PWManagedLocationManager *managedLocationManager = [[PWManagedLocationManager alloc] initWithBuildingId:self.buildingIdentifier];
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.mapView registerLocationManager:managedLocationManager];
    });
}

- (void)configureMapViewConstraints {
    self.mapView.translatesAutoresizingMaskIntoConstraints = NO;
    [[self.mapView.topAnchor constraintEqualToAnchor:self.view.topAnchor] setActive:YES];
    [[self.mapView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor] setActive:YES];
    [[self.mapView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor] setActive:YES];
    [[self.mapView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor] setActive:YES];
}

#pragma mark - PWMapViewDelegate

- (void)mapView:(PWMapView *)mapView locationManager:(id<PWLocationManager>)locationManager didUpdateIndoorUserLocation:(PWIndoorLocation *)userLocation {
    if (!self.firstLocationAcquired) {
        self.firstLocationAcquired = YES;
        self.mapView.trackingMode = PWTrackingModeFollow;
        
        NSInteger destinationPOIIdentifier = 0; /* Replace with the destination POI identifier */
        
        PWPointOfInterest *destinationPOI;
        if (destinationPOIIdentifier != 0) {
            NSPredicate *predicate = [NSPredicate predicateWithFormat:@"identifier = %d", destinationPOIIdentifier];
            NSArray *filteredArray = [self.mapView.building.pois filteredArrayUsingPredicate:predicate];
            destinationPOI = [filteredArray firstObject];
        } else {
            destinationPOI = [self.mapView.building.pois firstObject];
        }
        
        if (!destinationPOI) {
            NSLog(@"No points of interest found, please add at least one to the building in the Maas portal");
            return;
        }
        
        [PWRoute createRouteFrom:self.mapView.indoorUserLocation to:destinationPOI accessibility:NO excludedPoints:nil completion:^(PWRoute *route, NSError *error) {
            __weak typeof(self) weakSelf = self;
            
            if (route) {
                [weakSelf.mapView navigateWithRoute:route];
            }
        }];
    }
}

#pragma mark - CLLocationManagerDelegate

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status {
    if (status == kCLAuthorizationStatusAuthorizedAlways || status == kCLAuthorizationStatusAuthorizedWhenInUse) {
        [self startManagedLocationManager];
    } else {
        NSLog(@"Not authorized to start PWManagedLocationManager");
    }
}

@end

File: Samples/MapScenarios/MapScenarios/Scenarios/RoutingViewController.swift
//
//  RoutingViewController.swift
//  MapScenarios
//
//  Created on 3/7/18.
//  Copyright © 2018 Phunware. All rights reserved.
//

import Foundation
import UIKit
import PWCore
import PWMapKit

class RoutingViewController: UIViewController {
    
    // Enter your application identifier, access key, and signature key, found on Maas portal under Account > Apps
    var applicationId = ""
    var accessKey = ""
    var signatureKey = ""
    
    // Enter your building identifier here, found on the building's Edit page on Maas portal
    var buildingIdentifier = 0
    
    // Destination POI identifier for routing
    var destinationPOIIdentifier: Int = 0
    
    let mapView = PWMapView()
    let locationManager = CLLocationManager()
    var firstLocationAcquired = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = "Route to Point of Interest"
        
        if !validateBuildingSetting(appId: applicationId, accessKey: accessKey, signatureKey: signatureKey, buildingId: buildingIdentifier) {
            return
        }
        
        if destinationPOIIdentifier == 0 {
            warning("Please set a valid value for `destinationPOIIdentifier` in RoutingViewController.swift")
            return
        }
        
        PWCore.setApplicationID(applicationId, accessKey: accessKey, signatureKey: signatureKey)
        
        mapView.delegate = self
        view.addSubview(mapView)
        configureMapViewConstraints()
        
        // Start loading building
        PWBuilding.building(withIdentifier: buildingIdentifier) { [weak self] (building, error) in
            self?.mapView.setBuilding(building, animated: true, onCompletion: { (error) in
                if let error = error {
                    self?.warning(error.localizedDescription)
                    return
                }
                self?.locationManager.delegate = self
                if !CLLocationManager.isAuthorized() {
                    self?.locationManager.requestWhenInUseAuthorization()
                } else {
                    self?.startManagedLocationManager()
                }
            })
        }
    }
    
    func startManagedLocationManager() {
        DispatchQueue.main.async { [weak self] in
            if let buildingIdentifier = self?.buildingIdentifier {
                let managedLocationManager = PWManagedLocationManager(buildingId: buildingIdentifier)
                self?.mapView.register(managedLocationManager)
            }
        }
    }
    
    func configureMapViewConstraints() {
        mapView.translatesAutoresizingMaskIntoConstraints = false
        mapView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        mapView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
        mapView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
        mapView.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
    }
    
    //*************************************
    // Plot route on the map
    //*************************************
    func route() {
        // Set tracking mode to follow me
        mapView.trackingMode = .follow
        
        // Find the destination POI
        let destinationPOI = mapView.building.pois.filter({
            return $0.identifier == destinationPOIIdentifier
        }).first
        if destinationPOI == nil {
            warning("No points of interest found, please add at least one to the building in the Maas portal")
            return
        }
        
        // Calculate a route and plot on the map
        PWRoute.createRoute(from: mapView.indoorUserLocation,
                            to: destinationPOI,
                            options: nil,
                            completion: { [weak self] (route, error) in
            guard let route = route else {
                self?.warning("Couldn't find a route from you current location to the destination.")
                return
            }
            
            let routeOptions = PWRouteUIOptions()
            // routeOptions.routeStrokeColor = <#routeStrokeColor#>
            // routeOptions.directionFillColor = <#directionFillColor#>
            // routeOptions.directionStrokeColor = <#directionStrokeColor#>
            // routeOptions.instructionFillColor = <#instructionFillColor#>
            // routeOptions.instructionStrokeColor = <#instructionStrokeColor#>
            // routeOptions.showJoinPoint = <#true or false#>
            // routeOptions.joinPointColor = <#joinPointColor#>
            // routeOptions.lineJoin = <#.miter, round or bevel#>
            self?.mapView.navigate(with: route, options: routeOptions)
        })
    }
}

// MARK: - PWMapViewDelegate

extension RoutingViewController: PWMapViewDelegate {
    
    func mapView(_ mapView: PWMapView!, locationManager: PWLocationManager!, didUpdateIndoorUserLocation userLocation: PWUserLocation!) {
        if !firstLocationAcquired {
            firstLocationAcquired = true
            route()
        }
    }
}

// MARK: - CLLocationManagerDelegate

extension RoutingViewController: CLLocationManagerDelegate {
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        switch status {
        case .authorizedAlways, .authorizedWhenInUse:
            startManagedLocationManager()
        default:
            mapView.unregisterLocationManager()
            print("Not authorized to start PWLocationManager")
        }
    }
}
